File: tests/src/python/test_qgscomposereffects.py
# -*- coding: utf-8 -*-
"""QGIS Unit tests for QgsComposerEffects.

.. note:: This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
"""
__author__ = '(C) 2012 by Dr. Horst Düster / Dr. Marco Hugentobler'
__date__ = '20/08/2012'
__copyright__ = 'Copyright 2012, The QGIS Project'
# This will get replaced with a git SHA1 when you do a git archive
__revision__ = '$Format:%H$'

import qgis  # NOQA

from qgis.PyQt.QtGui import QPainter, QColor

from qgis.core import (QgsComposerShape,
                       QgsComposition,
                       QgsMapRenderer
                       )
from qgis.testing import start_app, unittest
from utilities import unitTestDataPath
from qgscompositionchecker import QgsCompositionChecker

start_app()
TEST_DATA_DIR = unitTestDataPath()


class TestQgsComposerEffects(unittest.TestCase):

    def __init__(self, methodName):
        """Run once on class initialization."""
        unittest.TestCase.__init__(self, methodName)

        # create composition
        self.mMapRenderer = QgsMapRenderer()
        self.mComposition = QgsComposition(self.mMapRenderer)
        self.mComposition.setPaperSize(297, 210)

        self.mComposerRect1 = QgsComposerShape(20, 20, 150, 100, self.mComposition)
        self.mComposerRect1.setShapeType(QgsComposerShape.Rectangle)
        self.mComposerRect1.setBackgroundColor(QColor.fromRgb(255, 150, 0))
        self.mComposition.addComposerShape(self.mComposerRect1)

        self.mComposerRect2 = QgsComposerShape(50, 50, 150, 100, self.mComposition)
        self.mComposerRect2.setShapeType(QgsComposerShape.Rectangle)
        self.mComposerRect2.setBackgroundColor(QColor.fromRgb(0, 100, 150))
        self.mComposition.addComposerShape(self.mComposerRect2)

    def testBlendModes(self):
        """Test that blend modes work for composer items."""

        self.mComposerRect2.setBlendMode(QPainter.CompositionMode_Multiply)

        checker = QgsCompositionChecker('composereffects_blend', self.mComposition)
        checker.setControlPathPrefix("composer_effects")
        myTestResult, myMessage = checker.testComposition()

        self.mComposerRect2.setBlendMode(QPainter.CompositionMode_SourceOver)

        assert myTestResult, myMessage

    def testTransparency(self):
        """Test that transparency works for composer items."""

        self.mComposerRect2.setTransparency(50)

        checker = QgsCompositionChecker('composereffects_transparency', self.mComposition)
        checker.setControlPathPrefix("composer_effects")
        myTestResult, myMessage = checker.testComposition()

        self.mComposerRect2.setTransparency(100)

        assert myTestResult, myMessage

if __name__ == '__main__':
    unittest.main()

File: tests/src/core/testqgscomposereffects.cpp
/***************************************************************************
                         testqgscomposereffects.cpp
                         ----------------------
    begin                : April 2013
    copyright            : (C) 2013 by Marco Hugentobler, Nyall Dawson
    email                : nyall dot dawson at gmail.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include "qgsapplication.h"
#include "qgscomposition.h"
#include "qgsmultirenderchecker.h"
#include "qgscomposershape.h"
#include "qgsproject.h"
#include <QObject>
#include "qgstest.h"
#include <QColor>
#include <QPainter>

class TestQgsComposerEffects : public QObject
{
    Q_OBJECT

  public:
    TestQgsComposerEffects() = default;

  private slots:
    void initTestCase();// will be called before the first testfunction is executed.
    void cleanupTestCase();// will be called after the last testfunction was executed.
    void init();// will be called before each testfunction is executed.
    void cleanup();// will be called after every testfunction.
    void blend_modes(); //test if composer item blending is functioning
    void transparency(); //test if composer transparency is functioning

  private:
    QgsComposition *mComposition = nullptr;
    QgsComposerShape *mComposerRect1 = nullptr;
    QgsComposerShape *mComposerRect2 = nullptr;
    QString mReport;
};

void TestQgsComposerEffects::initTestCase()
{
  QgsApplication::init();
  QgsApplication::initQgis();

  //create composition with two rectangles

  mComposition = new QgsComposition( QgsProject::instance() );
  mComposition->setPaperSize( 297, 210 ); //A4 landscape
  mComposerRect1 = new QgsComposerShape( 20, 20, 150, 100, mComposition );
  mComposerRect1->setShapeType( QgsComposerShape::Rectangle );
  mComposerRect1->setBackgroundColor( QColor::fromRgb( 255, 150, 0 ) );
  mComposition->addComposerShape( mComposerRect1 );
  mComposerRect2 = new QgsComposerShape( 50, 50, 150, 100, mComposition );
  mComposerRect2->setBackgroundColor( QColor::fromRgb( 0, 100, 150 ) );
  mComposerRect2->setShapeType( QgsComposerShape::Rectangle );
  mComposition->addComposerShape( mComposerRect2 );

  mReport = QStringLiteral( "<h1>Composer Effects Tests</h1>\n" );
}

void TestQgsComposerEffects::cleanupTestCase()
{
  delete mComposition;

  QString myReportFile = QDir::tempPath() + "/qgistest.html";
  QFile myFile( myReportFile );
  if ( myFile.open( QIODevice::WriteOnly | QIODevice::Append ) )
  {
    QTextStream myQTextStream( &myFile );
    myQTextStream << mReport;
    myFile.close();
  }

  QgsApplication::exitQgis();
}

void TestQgsComposerEffects::init()
{

}

void TestQgsComposerEffects::cleanup()
{

}

void TestQgsComposerEffects::blend_modes()
{
  mComposerRect2->setBlendMode( QPainter::CompositionMode_Multiply );

  QgsCompositionChecker checker( QStringLiteral( "composereffects_blend" ), mComposition );
  checker.setControlPathPrefix( QStringLiteral( "composer_effects" ) );
  QVERIFY( checker.testComposition( mReport ) );
  // reset blending
  mComposerRect2->setBlendMode( QPainter::CompositionMode_SourceOver );
}

void TestQgsComposerEffects::transparency()
{
  mComposerRect2->setItemOpacity( 0.5 );

  QgsCompositionChecker checker( QStringLiteral( "composereffects_transparency" ), mComposition );
  checker.setControlPathPrefix( QStringLiteral( "composer_effects" ) );
  QVERIFY( checker.testComposition( mReport ) );
}

QGSTEST_MAIN( TestQgsComposerEffects )
#include "testqgscomposereffects.moc"
